src/pubsub/core/callables.py

```diff
@@ -144,63 +149,73 @@ def __init__(self, func: UserListener, firstArgIdx: int, ignoreArgs: Sequence[st
             self.autoTopicArgName = None.
         """
 
-        (allParams, varParamName, varOptParamName, defaultVals) = getargspec(func)
-        if defaultVals is None:
-            defaultVals = []
-        else:
-            defaultVals = list(defaultVals)
+        args, varParamName, varOptParamName, argsDefaults, kwargs, kwargsDefaults, annotations = inspect.getfullargspec(func)
+        self.allArgs = {}
+
+        if (argsDefaults != None):
+            argsDefaults_startsAt = len(args) - len(argsDefaults) - 1
+        for i, variable in enumerate(args):
+            if ((i == 0) and (firstArgIdx > 0)):
+                continue #skip self
+
+            if ((argsDefaults == None) or (i < argsDefaults_startsAt)):
+                self.allArgs[variable] = self.NO_DEFAULT()
+            else:
+                self.allArgs[variable] = argsDefaults[i - argsDefaults_startsAt - 1]
+
+        self.allKwargs = {}
+        for variable in kwargs:
+            if ((kwargsDefaults == None) or (variable not in kwargsDefaults)):
+                self.allKwargs[variable] = self.NO_DEFAULT()
+            else:
+                self.allKwargs[variable] = kwargsDefaults[variable]
 
         self.acceptsAllKwargs = (varOptParamName is not None)
         self.acceptsAllUnnamedArgs = (varParamName is not None)
-
-        self.allParams = allParams
-        del self.allParams[0:firstArgIdx]  # does nothing if firstArgIdx == 0
+        self.allParams = [*self.allArgs.keys(), *self.allKwargs.keys()]
 
         if ignoreArgs:
             for var_name in ignoreArgs:
-                required = len(self.allParams) - len(defaultVals)
-                index = self.allParams.index(var_name)
-                del self.allParams[index]
-                if index >= required:
-                    del defaultVals[index - required]
+                if (var_name in self.allArgs):
+                    del self.allArgs[var_name]
+                elif (var_name in self.allKwargs):
+                    del self.allKwargs[var_name]
 
-            if varOptParamName in ignoreArgs:
+            if (varOptParamName in ignoreArgs):
                 self.acceptsAllKwargs = False
-            if varParamName in ignoreArgs:
+            if (varParamName in ignoreArgs):
                 self.acceptsAllUnnamedArgs = False
 
-        self.numRequired = len(self.allParams) - len(defaultVals)
+        self.numRequired = sum([1 for value in [*self.allArgs.values(), *self.allKwargs.values()] if (isinstance(value, self.NO_DEFAULT))])
```

Comment:
could `len()` and `filter()` be used to accomplish this more cleanly?

Reply:
:) I am all for things being done more cleanly.

Reply:
I tried doing what you suggested, but it looks less clean to me. Maybe you could take a crack at it?